Merge of 'feature/rdss-1948_versioning_clarifications' branch into 'develop' branch

danielgrant opened the pull request.

jhsimpson commented on README.md:
```diff
@@ -313,19 +313,25 @@ The following example Message payloads are provided in the [`messages/body/`](me
 
+### Message Identification
+
+Every dataset published by a producer **MUST** contain a unique `objectUuid` field. This field uniquely identifies a specific datatset under a specific version, and **MUST NOT** be duplicated when publishing subsequent versions of the same dataset.
```
the terms 'dataset' and 'digital object' appear to be used somewhat interchangeably in here?  This refers specifically to a message body conforming to the research object schema?  Or does this apply to other kinds of digital objects like material asset or intellectual asset?

It might be good to define a 'digital object' in the json schema section above, as any object defined by one of these schemas in the body of a message?

danielgrant replied:
Hi @jhsimpson, I have updated the structure and text to make it clear that this text applies specifically to digital objects within a metadata message.

mjaddis commented:
The Object Versioning section says: "A dataset MUST generate a new version when any file or property of the dataset is altered, and in such circumstances the Update version of the appropriate Message is used to distribute the updated payload."   but the Message Triggers section says: "A MetadataCreate Message MUST be sent when ...  A significant change to the work item is executed, such as changing the title or adding / removing files."   This is inconsistent.

mjaddis commented:
The Preservation Systems section says: "Upon receiving a MetadataCreate or MetadataUpdate payload, a PS MUST generate a preservation item which contains both the metadata contained within the payload and any files referenced by that metadata."    Should this be revised to say that the preservation item MUST contain any changed/additional files rather than all files, which allows for metadata only AIPs and other ways of dealing with updates to datasets that don't involve duplicating everything in the dataset between versions.

mjaddis commented:
In the Metadata Update section it says: "In the event that a Message is modified in such a way as to change its visibility from confidential to public, this should instead be treated as MetadataCreate as consumers of the Message SHOULD NOT have received a MetadataCreate Message for its original creation."  What happens if a dataset goes from Public->Private->Public?  Maybe a bit of an edge case edge case, but is that a series of MetadataUpdates, i.e. Private->Public results in a MetadataCreate only if a dataset has never been published before?

mjaddis commented:
There is nothing in the README that explicitly says that a MetadataCreate or MetadataUpdate for a revised dataset MUST contain a complete description of all the metadata and files for that dataset and MUST NOT not contain just the changes.  In other words, all MetadataCreate and MetadataUpdate messages MUST fully define the dataset without any dependency on information in previous messages.

mjaddis commented:
Is there some way to include some form of 'change log' in a MetadataUpdate message, e.g. which files have been added/deleted/changed in the dataset or which DataCite fields have been updated?  Otherwise the consuming application will in effect need to do a diff between the payloads of the current message and a previous message in order to work out what has changed.   This isn't always going to be simple, e.g. for files that are listed in the payload the order may or may not matter.

danielgrant commented:
Hi @mjaddis,

> The Object Versioning section says: "A dataset MUST generate a new version when any file or property of the dataset is altered, and in such circumstances the Update version of the appropriate Message is used to distribute the updated payload." but the Message Triggers section says: "A MetadataCreate Message MUST be sent when ... A significant change to the work item is executed, such as changing the title or adding / removing files." This is inconsistent.

Oversight on my part, I've fixed it now.

> The Preservation Systems section says: "Upon receiving a MetadataCreate or MetadataUpdate payload, a PS MUST generate a preservation item which contains both the metadata contained within the payload and any files referenced by that metadata." Should this be revised to say that the preservation item MUST contain any changed/additional files rather than all files, which allows for metadata only AIPs and other ways of dealing with updates to datasets that don't involve duplicating everything in the dataset between versions.

I have updated the text to reflect that it **MUST** contain the entire payload. We need to consider applications may see the object for the first time in a `MetadataUpdate` (for example, new systems onboarding or a system that was offline for a period of time). In this case, the system effectively treats it as a `MetadataCreate`, but would require a complete payload to do so.

> In the Metadata Update section it says: "In the event that a Message is modified in such a way as to change its visibility from confidential to public, this should instead be treated as MetadataCreate as consumers of the Message SHOULD NOT have received a MetadataCreate Message for its original creation." What happens if a dataset goes from Public->Private->Public? Maybe a bit of an edge case edge case, but is that a series of MetadataUpdates, i.e. Private->Public results in a MetadataCreate only if a dataset has never been published before?

This does sound like an edge case. I think sticking with the defined behaviour sounds like the best approach, whereby we keep it as a `MetadataUpdate`. Combined with the fact that every message (including `MetadataUpdate`'s) include all the fields (not just the changed ones), there's no risk of data loss etc here.

> There is nothing in the README that explicitly says that a MetadataCreate or MetadataUpdate for a revised dataset MUST contain a complete description of all the metadata and files for that dataset and MUST NOT not contain just the changes. In other words, all MetadataCreate and MetadataUpdate messages MUST fully define the dataset without any dependency on information in previous messages.

As above, I've updated this now to make it explicit that it **MUST** contain a complete description.

> Is there some way to include some form of 'change log' in a MetadataUpdate message, e.g. which files have been added/deleted/changed in the dataset or which DataCite fields have been updated? Otherwise the consuming application will in effect need to do a diff between the payloads of the current message and a previous message in order to work out what has changed. This isn't always going to be simple, e.g. for files that are listed in the payload the order may or may not matter.

I think there could be some value in this but it sounds non-trivial. @jhsimpson had suggested that Archivematica at least would just do a diff, and I'm inclined to stick with this for now. Any change to implement this would be a significant change and result in a new major version.

jhsimpson commented:
I didn't mean to suggest that doing a diff from Archivematica would be somehow easier than producing a changelog - it is likely more work for a consumer to sort out what has changed rather than the producer providing the diff.

That said, i agree that adding a changelog is a significant change to this spec, and could be left out of this Pull Request.  In practice, this is likely going to mean that consumers such as Archivematica will need to treat all metadataUpdates as significant changes, at least in the first implementation of this change.  I believe from discussions outside github that this is acceptable - we could revisit how best to produce a changelog/diff later.
